### ISSUE TITLE
Support running several commands more intuitively

### ISSUE THREAD

**Issue by jvican** [sbt/sbt#3051]
## problem
For new contributors, remembering that the sbt syntax to run several commands at once is `; myCommand; myCommand2` is hard and unintuitive.
  
Why is it unintuitive? Scala/Java programs don't follow this convention to execute several statements (**EDIT**: it could be the case that bash doesn't either, though there's a slight change in semantics).

## expectation
New users should run `myCommand; myCommand2`to execute sequentially `myCommand` and `myCommand2`. The current syntax can be deprecated or still allowed, but the new one should work out of the box.

In Scala (and also bash), `myCommand; myCommand2` is valid. This is why the proposed syntax is more intuitive and user-friendly.

**Comment by dwijnand** [comment]
The problem with this suggestion is that the semantics are different between sbt's prefix `;` notation vs bash's (POSIX's, I think) `;`:

## sbt

```scala
val foo = taskKey[Int]("")
val bar = taskKey[Int]("")

foo := (sys error "boom")
bar := 2
```

```
> ;foo ;bar
[trace] Stack trace suppressed: run last *:foo for the full output.
[error] (*:foo) boom
[error] Total time: 0 s, completed 28-Mar-2017 14:01:57
```

## bash

```bash
$ foo() { echo "boom" >&2; false; }
$ bar() { echo 2; }

$ foo; bar
boom
2

$ echo $?
0
```

## notes

bash's `&&` notation is more what sbt's `;` command does:

```bash
$ foo && bar
boom

! echo $?
1
```

**Comment by jvican** [comment]
True, but I think it's a fair limitation. Users are aware that throwing exceptions will always stop programs if you don't catch them.
  
bash's `&&` notation cannot be implemented in Scala because `&&` is not a supported statement delimiter in the Scala language. However, `;` is.
  
My suggestions focus on making this easier to guess and use. It's lightweight and familiar syntax. I'm happy to hear alternative proposals that have the same goals.

**Comment by milessabin** [comment]
I'm very much in favour or this, the semantic differences from POSIX notwithstanding.

I think this comes in at level "2. lexical syntax" in Wadler's law, BTW ;-)

**Comment by dwijnand** [comment]
sbt's shell isn't a POSIX shell, but it's not a good idea to make the syntax more similar while being different in semantics.

> bash's `&&` notation cannot be implemented in Scala because `&&` is not a supported statement delimiter in the Scala language. However, `;` is.
>
> My suggestions focus on making this easier to guess and use. It's lightweight and familiar syntax. I'm happy to hear alternative proposals that have the same goals.

Fortunately we don't need to implement it in Scala. We need to implement it in sbt shell.

My alternative proposal is `&&`.

**Comment by dwijnand** [comment]
As an aside, looks like you can implement `&&`, at least in some cases, in Scala:

```scala
scala> case class C(val x: Int) { def &&(n: Int) = x + n }
defined class C

scala> C(1) && 2
res0: Int = 3
```

**Comment by jvican** [comment]
Allowing `&&` is a bad idea because no one prevents you from defining a method that is `&&` and put it in infix position.
  
Those semantic differences you mention are not relevant. Java and Scala programs work this way. You cannot run in the same process two programs if the first one fails with an exception. This is something that all devs in the community would expect.

**Comment by jvican** [comment]
To make it more clear, what matters is that we're more similar to Scala. See `myCommand; myCommand2` as equivalent to:
  
```
myBoringTask := {
  myCommand
  myCommand2
}
```
  
and then calling `myBoringTask` from shell.
  
In this case, semantics would be the same and it's IMO what we should aim for :wink:.

**Comment by dwijnand** [comment]
> Those semantic differences you mention are not relevant.

I disagree with you.

> Java and Scala programs work this way. You cannot run in the same process two programs if the first one fails with an exception. This is something that all devs in the community would expect.

What about threads? What about process forking?

---

Btw, the semantics of POSIX's `;` are implemented in the `all` command:

```
> help all
all <task>+

	Executes all of the specified tasks concurrently.
```

**Comment by dwijnand** [comment]
> To make it more clear, what matters is that we're more similar to Scala.

No it doesn't. It's sbt shell. If it were "sbt REPL" then one could argue that. But as a shell we should match shell-like syntax, but then also match shell-like behaviour.

**Comment by jvican** [comment]
> Btw, the semantics of POSIX's ; are implemented in the all command.
  
No, POSIX's doesn't execute them concurrently.
  
> No it doesn't. It's sbt shell. If it were "sbt REPL" then one could argue that. But as a shell we should match shell-like syntax, but then also match shell-like behaviour.
  
We're not doing a good job at that, either. If your goal is to match shell-like behaviour, then we can change the implementation of `;command1;command2` to match POSIX's.
  
> I disagree with you.
  
Can you please specify why you think are relevant? The point I'm trying to make is that people expect these semantics as they match Java/Scala semantics. For me, sbt is more about Scala than the shell. I think others would agree with this statement. There's nothing similar to POSIX in sbt aside from `exit`. The similarity between the two in terms of semantics is anecdotal.
  
> What about threads? What about process forking?
  
Why are we talking about these? `;` syntax only allows sequential execution.
  
In general, my point is that this is the only reasonable and safe way to make this syntax easier to sbt users. Your suggestion of using `&&` is not possible. If the problem is semantics, we can change the underlying implementation. But I'm pretty sure this is not necessary. Users expect this behaviour -- exceptions work this way.

**Comment by dwijnand** [comment]
> No, POSIX's doesn't execute them concurrently.

Doh! You're right. ✅

---

> We're not doing a good job at that, either. If your goal is to match shell-like behaviour, then we can change the implementation of `;command1;command2` to match POSIX's.

I agree we're not. This ticket is about making sbt shell more similar to POSIX. As you say, make it "more intuitive and user-friendly".

---

> Why are we talking about these? `;` syntax only allows sequential execution.

With threads and process forking you can run "two programs if the first one fails with an exception". But you don't even need that, you can just wrap it with `try`.

Btw, looks like you can't do that in bash with `;` in one command:

```bash
$ ( sleep 1; echo 1) & ; echo 2
-bash: syntax error near unexpected token `;'
```

--- 

> Your suggestion of using `&&` is not possible.

Could you explain why you think `&&` is not possible?

---

For the rest I guess we're going to have to agree to disagree on what we think "users expect" and what "others would agree with".

**Comment by jvican** [comment]
> Btw, looks like you can't do that in bash with ; in one command:
  
This is working for me in Linux.
  
> Could you explain why you think && is not possible?
  
Yes, it's not possible because the syntax can be overriden by a user-defined method `&&`. Preventing users from doing so is bad, you're messing with Scala syntax. Also, once you implement `&&`, you'll need to implement `||` and friends.
  
My solution is simpler. If you really care about semantics, it can be accomodated by changing the underlying implementation of `;`.
  
@dwijnand We do disagree on what's expected. But I hope we agree that the status quo is undesirable and confusing.

**Comment by jvican** [comment]
> This ticket is about making sbt shell more similar to POSIX. As you say, make it "more intuitive and user-friendly".
  
By the way, I've removed this from the description. This ticket is about making it more intuitive by following Scala/Java semantics.

**Comment by dwijnand** [comment]
> Yes, it's not possible because the syntax can be overriden by a user-defined method `&&`. Preventing users from doing so is bad, you're messing with Scala syntax. Also, once you implement `&&`, you'll need to implement `||` and friends.

You're confused. This isn't Scala syntax. This is sbt shell. So it's either sbt's command parsers. Or perhaps it could be changed/fixed where sbt shell passes the user input to the command engine. Or maybe where sbt passes the command input to the task engine. Either way we're not talking about Scala syntax here.

> @dwijnand We do disagree on what's expected. But I hope we agree that the status quo is undesirable and confusing.

I'm used to it, but I'm happy to change it. Btw looks like you still state "This is why the proposed syntax is more intuitive and user-friendly."

**Comment by jvican** [comment]
> I'm used to it, but I'm happy to change it. Btw looks like you still state "This is why the proposed syntax is more intuitive and user-friendly."
  
I'm not proposing to remove it, it's fine to leave it as soon as the new proposed syntax works.
  
>  So it's either sbt's command parsers.
  
You're right. Then I propose two things:
  
1. Add `&&` and `||`.
2. Add my proposed syntax `foo; bar` as a shorcut for `;foo; bar` and match POSIX behaviour.
  
Those two things would satisfy people looking for the different semantics we're discussing.

**Comment by dwijnand** [comment]
What do you mean by "match POSIX behaviour"? Do you mean change `;foo ;bar`'s behaviour?

---

I propose we do the following (and I propose we track it in separate issues):

## Add infix `&&`

Example:

```
> clean && compile
```

Equivalent in behaviour to the current

```
> ;clean ;compile
```

That is it runs sequentially, and stops at the first failure.

## Add suffix `;` syntax

Example:

```
> clean; compile;
> clean; compile
```

Similar to the current

```
> ;clean ;compile
```

except it doesn't stop at a failure.

**Comment by jvican** [comment]
> What do you mean by "match POSIX behaviour"? Do you mean change ;foo ;bar's behaviour?
  
Yes, I mean that.
  
> I propose we do the following (and I propose we track it in separate issues)
  
As you prefer, I don't mind. I'll open those issues.

**Comment by milessabin** [comment]
I'm struggling to understand your objections @dwijnand ... are there _any_ circumstances in which you wouldn't want to stop on failure? If there are, are they common enough to be considered the default case (as opposed to something which should be handled by something like `||`)?

**Comment by dwijnand** [comment]
Oh! Thank you Jorge for creating the tickets.

> are there any circumstances in which you wouldn't want to stop on failure? If there are, are they common enough to be considered the default case

I can imagine it would be useful to be able to do something like:

```scala
> foo/test; bar/test; baz/test
```

when wanting to run the tests of all 3 projects without stopping at the first one that fails.

It might not be a very common circumstance, but what do you mean by "the default case"? The `;foo; bar` syntax isn't going anywhere or changing in behaviour.

I think we'd do best to follow POSIX's precedent here, rather than match its syntax but with different behaviour.

**Comment by milessabin** [comment]
I mean that if I type (as I do several times a day) `clean ; compile` it does what everyone expects it to, clean, then compile, stopping on failure. I don't mind if `; clean ; compile` continues to do what it does today, but think the proposal in this issue is spot on.

Your example of running multiple sets of tests in parallel is reasonable, but I think atypical. I would like to see it supported by an alternative syntax, for instance `foo/test || bar/test || baz/test`.

We should be optimizing for the common case, not the exceptions.

**Comment by dwijnand** [comment]
My example would run them sequentially by the way. If you want to run tasks in parallel, that's what the `all` command does.

The syntax of `||` in POSIX runs the RHS if the LHS fails:

```bash
$ false || echo hi
hi

$ true || echo hi

$
```

I'm happy to optimise for the common case, but isn't `clean && compile` just as optimised? It _is_ 1 character longer, but I think that's a good trade-off to match POSIX.

**Comment by eed3si9n** [comment]
In my opinion `;foo ; bar;` is an example of sbt being too cute/clever about the implementation.

Similar to `-` being a *command* (!) for early processing in sbt 0.13 (we are renaming that in sbt 1.0 https://github.com/sbt/sbt/pull/2741), `;` is implemented as a *command* that does sequential processing. See https://github.com/sbt/sbt/blob/4e267e5cb5c805ae3400f5b22c853601f612ccce/main-command/src/main/scala/sbt/BasicCommands.scala#L84-L93 and https://github.com/sbt/sbt/blob/4e267e5cb5c805ae3400f5b22c853601f612ccce/main-command/src/main/scala/sbt/BasicCommandStrings.scala#L118-L123 for more details:

```scala
  def Multi = ";"
```

By using this strategy sbt shell implements `;` on the user-land, similar to Bash implementing some syntactical construct as shell script. Even though I've come to internalize this over the years, I agree with @jvican's problem statement that it's confusing to new users.

I think we should treat `;` as a proper syntactical construct. There's a potential risk that come command may have used `;` for different semantics, but that's likely a moot argument since we already have to deal with with `; foo; bar a; b`. (Here assuming `bar` command uses `a; b` as argument)

**Comment by jvican** [comment]
> I'm happy to optimise for the common case, but isn't clean && compile just as optimised? It is 1 character longer, but I think that's a good trade-off to match POSIX.
  
The problem, IMO, is that this introduces extra syntax that people have to learn and remember.
  
As Scala/Java programmers, the first thing that comes to mind when we want to run more than one statement is to separate them with `;`. It's people's first guess.
  
If we're to optimize for the common use case, I think dropping the leading `;` is the best thing to do.
  
> I think we should treat ; as a proper syntactical construct. There's a potential risk that come command may have used ; for different semantics, but that's likely a moot argument since we already have to deal with with ; foo; bar a; b. (Here assuming bar command uses a; b as argument)
  
Agreed.
  
> Even though I've come to internalize this over the years, I agree with @jvican's problem statement that it's confusing to new users.
  
I remember the first time I learned about this after 15 minutes looking up in the Internet how to do it. It was confusing. Now, after several years, it feels natural to append the leading `;`. But I think it's important that we try to forget what we know as power users, and optimize for the benefit of new users so that sbt's learning curve is not as steep as it is. I am happy you agree this is a problem.

**Comment by eed3si9n** [comment]
> As Scala/Java programmers, the first thing that comes to mind when we want to run more than one statement is to separate them with ;. It's people's first guess.

If we're saying that the semantics of `;` is that of procedural statements like Scala/Java, then it *should* fail to publish in `;clean; test; publishSigned` if `test` fails.

#3060

**Comment by jvican** [comment]
Yes, which are the concrete semantics of the `;` command. I created those tickets per @dwijnand's suggestion, but I still think that the solution proposed in this ticket is the best one for beginners.
  
It looks like a good candidate for 1.0.

**Comment by dwijnand** [comment]
> The problem, IMO, is that this introduces extra syntax that people have to learn and remember.
>
> As Scala/Java programmers, the first thing that comes to mind when we want to run more than one statement is to separate them with `;`. It's people's first guess.

I believe you've correctly identified here where our assumptions differ. You're right that as Scala/Java programmers they would reach for `;`. But as shell users if they want stop-on-failure they would reach for `&&`.

But I've changed my mind because of something that Eugene said in #3060:

> users are far more likely to expect stop-on-failure for `;clean; test; publishSigned` whether the semicolon is located on prefix or postfix.

Having semantics be so different based only on the presence or absence of a leading semicolon is far too subtle. Put differently: it's too late to try and match POSIX in sbt shell.

This means that #3060 no longer makes sense. And #3061 isn't necessary. I'll close them both.

**Comment by nadavwr** [comment]
Reading through this thread, I now understand from @eed3si9n's explanation of how `;` is implements—as John Oliver puts it—"how is that still a thing".

I understand that postfix-semicolon is a somewhat larger undertaking than one would initially imagine, and might not be justified in order to replace an odd piece of syntax that people have gotten used to in the past and should continue to get used to with little issue in the future.

However, I do find the entire strand of this discussion pitting SBT `;` semantics against POSIX to be a bit of a stretch—all due respect to POSIX, it is common practice start off build and installation scripts with a `set -o errexit` (aka `set -e`) because that is what users expect within the context of a build.

**Comment by jvican** [comment]
> I understand that postfix-semicolon is a somewhat larger undertaking than one would initially imagine
  
It doesn't matter. It's an annoying syntax that puts beginners off. This is part of all those small details that make people say that "sbt has a steep learning curve".
  
> an odd piece of syntax that people have gotten used to in the past and should continue to get used to with little issue in the future.
  
I think we could keep the semantics of the previous syntax, or enable it under a flag.

**Comment by eed3si9n** [comment]
> I understand that postfix-semicolon is a somewhat larger undertaking than one would initially imagine

Maybe I'm optimistic, but I feel like this is actually on the lower hanging side among the nice-to-have fruits are concerned.

**Comment by nadavwr** [comment]
@jvican I am entirely in favor of adopting your proposal. Reading through, I just realized why SBT maintainers might think the effort unjustified.

If this feature is to remain implemented as a command, then how about adding a `(`, which will behave almost exactly like `;`, with the added constraint that a matching number of `)` would be required?

This would leave the existing `;command1;command2` usage as it was, but would also allow `(command1;command2)` with identical semantics. This syntax will cause no eyebrows to be raised.

This of course can be substituted with any other form of bracketing if parens aren't available (e.g. if there's some clash with `early()`).

*edit:* Delimeters can't be used as command names. If this were to be relaxed a bit...

**Comment by eed3si9n** [comment]
@eatkins just sent in https://github.com/sbt/sbt/pull/4456, so this might be fixed in 1.3.0.
